ENH: Add cloudwatch monitoring

greglucas opened the pull request:
# Change Summary

## Overview
This adds a slack integration to send cloudwatch alarms to via an SNS Topic. The slack integration required a manual setup due to authorization permissions, but it wasn't too onerous and the steps are documented in the docstring.

One question I have is whether we want to do it this way and pass in things we want monitored (i.e. we are passing in APIGateway). Or do we want to go the other way and pass our SNS Topic and set up the cloudwatch alarms/metrics everywhere we want to integrate them so they are closer to what we are monitoring?

Is Slack integration OK for now, or do we also want to add e-mail notification? That is easy enough to add by just subscribing an e-mail to the topic as well.

closes #160
closes #173

Example of it deployed:
![image](https://github.com/IMAP-Science-Operations-Center/sds-data-manager/assets/12417828/a22ab04e-024d-4a8c-9a36-727e955700cc)
And what it looks like in Slack:
<img width="645" alt="image" src="https://github.com/IMAP-Science-Operations-Center/sds-data-manager/assets/12417828/97ca6d20-c2ad-481d-92ee-5533dddca339">


## New Dependencies
<!--List any new dependencies introduced by these changes-->

## New Files
<!--List each new file and add bullets to describe the purpose/function of the new file.
This should be high level, but enough detail for the reviewer to understand what purpose(s) the
code in the file is serving-->
- monitoring_stack.py
   - Adds an SNS Topic for Slack integration and a cloudwatch alarm based on the APIGateway.

## Updated Files
<!--List each updated file and add bullets to describe the purpose/function of the updates.
This should be high level, but enough detail for the reviewer to understand what purpose(s) the
updated code in the file is serving-->
- api_gateway
   - Needed to output the api name (I couldn't figure out how to get this as a wildcard unfortunately)

greglucas committed:
ENH: Add cloudwatch monitoring

This adds a slack integration to send cloudwatch alarms to via
an SNS Topic. The slack integration required a manual setup
due to authorization permissions, but it wasn't too onerous
and the steps are documented in the docstring.

greglucas committed:
MNT: Rearrange monitoring to top level that other stacks can depend on

This moves the monitoring stack to the highest level that other
stacks can now depend on rather than being integrated into the end.
This will hopefully limit the number of stacks that need to be
re-deployed when things get updated.

greglucas commented:
I reworked this to put monitoring at the top now and pass the sns topic to a method called on the APIGateway stack now.

bourque commented on sds_data_manager/stacks/api_gateway_stack.py:
```diff
@@ -1,5 +1,7 @@
-from aws_cdk import Stack
```
I notice that a fair number of the modules in this repo don't have module docstrings. I try to make it a habit to include these, but perhaps I am in the minority. Do we think it is important to include module docstrings?

greglucas replied:
I agree, this repo is in a pretty inconsistent state right now :-/ I am +0.5 on module docstrings. So I added them in. I think in most of these cases it actually made sense to move the class docstrings up to the module level, so that is what I did.

tech3371 replied:
I have notice that and have been wondering if we should start doing that. I like both ways. Not particular about it.

bourque commented on sds_data_manager/stacks/api_gateway_stack.py:
```diff
@@ -86,7 +88,40 @@ def __init__(
+        )
+
+        # Define the alarm
+        cw_alarm = cw.Alarm(
```
```suggestion
        cloudwatch_alarm = cw.Alarm(
```
Unnecessarily abbreviated variable names is a pet peeve of mine, but maybe this one gets a pass if `cw` is just the standard way to name things like this?

greglucas replied:
Agreed, thanks for bringing this up. I updated this and the imports. I think we could actually remove a bunch of the `from aws_cdk import aws_xxx as xxx` and just change them to `from aws_cdk import aws_xxx`, the `aws_xxx` is actually somewhat helpful for disambiguating resources we create vs import from cdk. But, I'll leave that for a follow-up cleanup PR.

bourque commented on sds_data_manager/stacks/api_gateway_stack.py:
```diff
@@ -86,7 +88,40 @@ def __init__(
             # Create a new method that is linked to the Lambda function
             resource.add_method(http_method, apigw.LambdaIntegration(lambda_fn))
+
+    def subscribe_to_sns(self, sns_topic: aws_sns.Topic):
```
Missing docstring here

greglucas replied:
Thanks! Fixed.

bourque commented on sds_data_manager/stacks/monitoring_stack.py:
```diff
@@ -0,0 +1,26 @@
+
+    Set up an integration with Slack to send notifications to a channel.
+
+    This had to be done manually in the AWS console to authorize the
+    AWS account to use the Slack app.
+    Just follow the steps here to create a new Slack authorization.
+    https://us-east-2.console.aws.amazon.com/chatbot/home
+    Choosing the channel and permissions as you go through the steps.
```
Thanks for adding these instructions here!

greglucas replied:
Moved this info up to the module level now.

bourque reviewed (commented):
This looks good to me, nice work! I just had some docstring-related comments.

I don't think I know enough about what is going on / I don't have strong enough opinions to comment on your first question, but in regards to email notifications -- My personal preference would be to just stick with Slack alerts. I tend to instantly ignore automated emails 😅

greglucas committed:
DOC: Cleaning up docstrings for monitoring and apigateway

Missing some module level docstrings, and the naming conventions
of resources was not consistent.

tech3371 commented on sds_data_manager/stacks/monitoring_stack.py:
```diff
@@ -0,0 +1,43 @@
+        super().__init__(scope, construct_id, **kwargs)
+        self.sns_topic_notifications = aws_sns.Topic(
+            self, "sns-notifications", display_name="sns-notifications"
+        )
```
I don't any subscription to the `SNS`. Is it because we have to integrate to Slack manually?

greglucas replied:
Good catch! I actually hadn't deployed this new version when you commented because I was waiting until people voted on which method they'd prefer. I just deployed it though. But you are correct, you have to manually go in and subscribe the chatbot to this Topic. Do you want to do that now to see what it looks like?

tech3371 replied:
I see. We can add more subscription in future PR too

greglucas replied:
Exactly.

tech3371 reviewed (approved):
Looks good to me. It was easy to follow.

greglucas committed:
Merge branch 'dev' into cloudwatch-metrics
